### DISCUSSION TITLE
Write blob from R package?

### DISCUSSION THREAD

**Discussion by kaneplusplus** [duckdb/duckdb#3229] [General]
I've been enjoying this package very much. Thanks for all of the work.

Is there a way to write a blob from an R package? Ideally, it would look like the following.

```r
library(tibble)
library(duckdb)
library(blob)

con <- dbConnect(duckdb::duckdb(), dbdir=":memory:", read_only=FALSE)

x <- tibble(name = "mtcars", dataset = blob(serialize(mtcars, NULL)))

dbWriteTable(con, "datasets", x)
```

**Comment by hannes** [comment]
This seems to work for the retrieving side already:
````R
dbGetQuery(con, "SELECT 'asdf'::BLOB")
````

But indeed the resulting data frame can currently not be written back. We should probably fix that.

**Comment by kaneplusplus** [comment]
Got it. Thanks for the example.

As a temporary workaround, it's easy enough to write `raw` vectors to character and then convert back with the code below.

```r

library(tibble)
library(duckdb)
library(DBI)
library(dplyr)

raw_to_rawchar <- function(raw_vec) {
  paste(raw_vec, collapse = "")
}

rawchar_to_raw <- function(rawchar_vec) {
  ind <- matrix(seq_len(nchar(rawchar_vec)), ncol = 2, byrow = TRUE)
  ret <- vapply(
    seq_len(nrow(ind)),
    function(i) substr(rawchar_vec, ind[i,1], ind[i,2]),
    NA_character_
  )
  as.raw(as.hexmode(ret))
}

con <- dbConnect(duckdb::duckdb(), dbdir=":memory:", read_only=FALSE)

x <- tibble(name = "mtcars", dataset = raw_to_rawchar(serialize(mtcars, NULL)))

dbWriteTable(con, "datasets", x)

datasets <- tbl(con, "datasets") |> as_tibble()

datasets$dataset[[1]] |>
  rawchar_to_raw() |>
  unserialize() |>
  head()
```

**Comment by hannes** [comment]
So the issue is that a `raw` column in a `data.frame` is represented as a `list`. However, `list` does not require nor allows to enforce that every entry has the same type, so we would have to essentially check every entry whether its either `raw` or `NA`. What almost works is to go through Arrow:
````R
df <- data.frame(b=I(list(charToRaw('asdf'))))
at <- arrow::Table$create(df)
duckdb::duckdb_register_arrow(con, "at", at)
````

But alas

````R
str(dbGetQuery(con, "SELECT b FROM at"))
'data.frame':	1 obs. of  1 variable:
 $ b:List of 1
  ..$ : int  97 115 100 102
````

and 

````R
 dbGetQuery(con, "SELECT b::BLOB FROM at")
Error: rapi_execute: Failed to run query
Error: Conversion Error: Unimplemented type for cast (UTINYINT[] -> BLOB)
````

Maybe we should add that `CAST`, too :)

**Reply by kaneplusplus** [reply]
Interesting. So support is close. It looks like it's easy enough to get the raw values from the arrow object.

```r
> as.raw(at$b$as_vector())
[1] 61 73 64 66
> charToRaw("asdf")
[1] 61 73 64 66
```

As you mention, the difficulty is that the raw column needs to be stored as a list. This is a little bit strange since it seems like a `raw` should be stored an accessed as contiguous memory, more like an element of the `character` type. This probably the motivation for the `blob` library behavior, which I wasn't familiar with until today.

```r
library(blob)
> as_blob(list("3", 3))
Error: `x` must be a list of raw vectors
> as_blob(list(charToRaw("3"), serialize(3, NULL)))
<blob[2]>
[1] blob[1 B]  blob[39 B]
```

In any case, I really like this package. I hope the discussion is helpful. I have a fix for now and I'll follow for updates. 

Thanks very much.

**Comment by hannes** [comment]
The `blob` package looks nice, now if it just did not depend on `rlang` and `vectrs` we might actually be able to use it :/
